### ISSUE TITLE
Start With Gradient Boosting Results from Comparing 13 Algorithms on 165 Datasets

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#497]
Start With Gradient Boosting, Results from Comparing 13 Algorithms on 165 Datasets<br>
<br>
https://ift.tt/2J4C7mP<br>
<br>
Jason Brownlee<br>
<br>
<p><em><strong>Which machine learning algorithm should you use?</strong></em></p>
<p>It is a central question in applied machine learning.</p>
<p>In a recent paper by Randal Olson and others, they attempt to answer it and give you a guide for algorithms and parameters to try on your problem first, before spot checking a broader suite of algorithms.</p>
<p>In this post, you will discover a study and findings from evaluating many machine learning algorithms across a large number of machine learning datasets and the recommendations made from this study.</p>
<p>After reading this post, you will know:</p>
<ul>
<li>That ensemble tree algorithms perform well across a wide range of datasets.</li>
<li>That it is critical to test a suite of algorithms on a problem as there is no silver bullet algorithm.</li>
<li>That it is critical to test a suite of configurations for a given algorithm as it can result in as much as a 50% improvement on some problems.</li>
</ul>
<p>Let&rsquo;s get started.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/03/Start-With-Gradient-Boosting-but-Always-Spot-Check-Algorithms-and-Configurations.jpg" alt="Start With Gradient Boosting, but Always Spot Check Algorithms and Configurations" width="640" class="wp-image-5054" height="427">
<p class="wp-caption-text">Start With Gradient Boosting, but Always Spot Check Algorithms and Configurations<br>
Photo by <a href="https://www.flickr.com/photos/_rmt_/16132591083/">Ritesh Man Tamrakar</a>, some rights reserved.</p>
</div>
<h2>The Paper</h2>
<p>In 2017, <a href="http://www.randalolson.com/">Randal Olson</a>, et al. released a pre-print for a paper with the intriguing title &ldquo;<a href="https://arxiv.org/abs/1708.05070">Data-driven Advice for Applying Machine Learning to Bioinformatics Problems</a>&ldquo;.</p>
<p>The goal of their work was to address the question that every practitioner faces when getting started on their predictive modeling problem; namely:</p>
<p><em><strong>What algorithm should I use?</strong></em></p>
<p>The authors describe this problem as choice overload, as follows:</p>
<blockquote>
<p>Although having several readily-available ML algorithm implementations is advantageous to bioinformatics researchers seeking to move beyond simple statistics, many researchers experience &ldquo;choice overload&rdquo; and find difficulty in selecting the right ML algorithm for their problem at hand.</p>
</blockquote>
<p>They approach the problem by running a decent sample of algorithms across a large sample of standard machine learning datasets to see what algorithms and parameters work best in general.</p>
<p>They describe their paper as:</p>
<blockquote>
<p>&hellip; a thorough analysis of 13 state-of-the-art, commonly used machine learning algorithms on a set of 165 publicly available classification problems in order to provide data-driven algorithm recommendations to current researchers</p>
</blockquote>
<p>It is very much similar to the paper &ldquo;<a href="http://jmlr.csail.mit.edu/papers/v15/delgado14a.html">Do We Need Hundreds of Classifiers to Solve Real World Classification Problems?</a>&rdquo; covered in the post &ldquo;<a href="https://machinelearningmastery.com/use-random-forest-testing-179-classifiers-121-datasets/">Use Random Forest: Testing 179 Classifiers on 121 Datasets</a>&rdquo; both in methodology and findings.</p>
<h2>Machine Learning Algorithms</h2>
<p>A total of 13 different algorithms were chosen for the study.</p>
<p>Algorithms were chosen to provide a mix of types or underlying assumptions.</p>
<p>The goal was to represent the most common classes of algorithms used in the literature, as well as recent state-of-the-art algorithms</p>
<p>The complete list of algorithms is provided below.</p>
<ul>
<li>Gaussian Naive Bayes (GNB)</li>
<li>Bernoulli Naive Bayes (BNB)</li>
<li>Multinomial Naive Bayes (MNB)</li>
<li>Logistic Regression (LR)</li>
<li>Stochastic Gradient Descent (SGD)</li>
<li>Passive Aggressive Classifier (PAC)</li>
<li>Support Vector Classifier (SVC)</li>
<li>K-Nearest Neighbor (KNN)</li>
<li>Decision Tree (DT)</li>
<li>Random Forest (RF)</li>
<li>Extra Trees Classifier (ERF)</li>
<li>AdaBoost (AB)</li>
<li>Gradient Tree Boosting (GTB)</li>
</ul>
<p>The scikit-learn library was used for the implementations of these algorithms.</p>
<blockquote>
<p>Each algorithm has zero or more parameters, and a grid search across sensible parameter values was performed for each algorithm.</p>
</blockquote>
<p>For each algorithm, the hyperparameters were tuned using a fixed grid search.</p>
<p>A table of algorithms and the hyperparameters evaluated is listed below, taken from the paper.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/01/Table-of-Algorithms-and-Parameters-695x1024.png" alt="Table of Algorithms and Parameters" width="695" class="wp-image-5050" height="1024">
<p class="wp-caption-text">Table of Algorithms and Parameters</p>
</div>
<p>Algorithms were evaluated using 10-fold cross-validation and the balanced accuracy measure.</p>
<p>Cross-validation was not repeated, perhaps introducing some statistical noise into the results.</p>
<h2>Machine Learning Datasets</h2>
<p>A selection of 165 standard machine learning problems were selected for the study.</p>
<p>Many of the problems were drawn from the field of bioinformatics, although not all datasets belong to this field of study.</p>
<p>All prediction problems were classification type problems with two or more classes.</p>
<blockquote>
<p>The algorithms were compared on 165 supervised classification datasets from the Penn Machine Learning Benchmark (PMLB). [&hellip;] PMLB is a collection of publicly available classification problems that have been standardized to the same format and collected in a central location with easy access via Python.</p>
</blockquote>
<p>The datasets were drawn from the Penn Machine Learning Benchmark (PMLB) collection, which is a project that provides standard machine learning datasets in a uniform format and made available by a simple Python API. You can learn more about this dataset catalog on the GitHub Project:</p>
<ul>
<li><a href="https://github.com/EpistasisLab/penn-ml-benchmarks">Penn Machine Learning Benchmarks</a></li>
</ul>
<p>All datasets were standardized prior to fitting models.</p>
<blockquote>
<p>Prior to evaluating each ML algorithm, we scaled the features of every dataset by subtracting the mean and scaling the features to unit variance.</p>
</blockquote>
<p>Other data preparation was not performed, nor feature selection or feature engineering.</p>
<h2>Analysis of Results</h2>
<p>The large number of experiments performed resulted in a lot of skill scores to analyze.</p>
<p>The analysis of the results was handled well, asking interesting questions and providing findings in the form of easy-to-understand charts.</p>
<blockquote>
<p>The entire experimental design consisted of over 5.5 million ML algorithm and parameter evaluations in total, resulting in a rich set of data that is analyzed from several viewpoints&hellip;</p>
</blockquote>
<p>Algorithm performance was ranked for each dataset, then the average rank of each algorithm was calculated.</p>
<p>This provided a rough and easy to understand idea of which algorithms performed well or not, on average.</p>
<p>The results showed that both Gradient boosting and random forest had the lowest rank (performed best) and that the Naive Bayes approaches had the highest rank (performed worst) on average.</p>
<blockquote>
<p>The post-hoc test underlines the impressive performance of Gradient Tree Boosting, which significantly outperforms every algorithm except Random Forest at the p &lt; 0.01 level.</p>
</blockquote>
<p>This is demonstrated with a nice chart, taken from the paper.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/01/Algorithm-Mean-Rank-1024x481.png" alt="Algorithm Mean Rank" width="1024" class="wp-image-5051" height="481">
<p class="wp-caption-text">Algorithm Mean Rank</p>
</div>
<p>No single algorithm performs best or worst.</p>
<p>This is wisdom known to machine learning practitioners, but difficult to grasp for beginners in the field.</p>
<p>There is no silver bullet and you must test a suite of algorithms on a given dataset to see what works best.</p>
<blockquote>
<p>&hellip; it is worth noting that no one ML algorithm performs best across all 165 datasets. For example, there are 9 datasets for which Multinomial NB performs as well as or better than Gradient Tree Boosting, despite being the overall worst- and best-ranked algorithms, respectively. Therefore, it is still important to consider different ML algorithms when applying ML to new datasets.</p>
</blockquote>
<p>Further, picking the right algorithm is not enough. You must also pick the right configuration of the algorithm for your dataset.</p>
<blockquote>
<p>&hellip; both selecting the right ML algorithm and tuning its parameters is vitally important for most problems.</p>
</blockquote>
<p>The results found that tuning an algorithm lifted skill of a method anywhere from 3% to 50%, depending on the algorithm and the dataset.</p>
<blockquote>
<p>The results demonstrate why it is unwise to use default ML algorithm hyperparameters: tuning often improves an algorithm&rsquo;s accuracy by 3-5%, depending on the algorithm. In some cases, parameter tuning led to CV accuracy improvements of 50%.</p>
</blockquote>
<p>This is demonstrated with a chart from the paper that shows the spread of improvement offered by parameter tuning on each algorithm.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/01/Algorithm-performance-improvement-via-parameter-tuning-1016x1024.png" alt="Algorithm performance improvement via parameter tuning" width="1016" class="wp-image-5052" height="1024">
<p class="wp-caption-text">Algorithm performance improvement via parameter tuning</p>
</div>
<p>Not all algorithms are required.</p>
<p>The results found that five algorithms and specific parameters achieved top 1% in performance across 106 of the 165 tested datasets.</p>
<p>These five algorithms are recommended as a starting point for spot checking algorithms on a given dataset in bioinformatics, but I would suggest also more generally:</p>
<ul>
<li>Gradient Boosting</li>
<li>Random Forest</li>
<li>Support Vector Classifier</li>
<li>Extra Trees</li>
<li>Logistic Regression</li>
</ul>
<p>The paper provides a table of these algorithms, including the recommend parameter settings and the number of datasets covered, e.g. where the algorithm and configuration achieved top 1% performance.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/01/Suggested-Algorithms-1024x775.png" alt="Suggested Algorithms" width="1024" class="wp-image-5053" height="775">
<p class="wp-caption-text">Suggested Algorithms</p>
</div>
<h2>Practical Findings</h2>
<p>There are two big findings from this paper that are valuable for practitioners, especially those starting out or those who are under pressure to get a result on their own predictive modeling problem.</p>
<h3>1. Use Ensemble Trees</h3>
<p>If in doubt or under time pressure, use ensemble tree algorithms such as gradient boosting and random forest on your dataset.</p>
<blockquote>
<p>The analysis demonstrates the strength of state-of-the-art, tree-based ensemble algorithms, while also showing the problem-dependent nature of ML algorithm performance.</p>
</blockquote>
<h3>2. Spot Check and Tune</h3>
<p>No one can look at your problem and tell you what algorithm to use, and there is no silver bullet algorithm.</p>
<p>You must test a suite of algorithms and a suite of parameters for each algorithm to see what works best for your specific problem.</p>
<blockquote>
<p>In addition, the analysis shows that selecting the right ML algorithm and thoroughly tuning its parameters can lead to a significant improvement in predictive accuracy on most problems, and is a critical step in every ML application.</p>
</blockquote>
<p>I talk about this all the time; for example, see the post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/a-data-driven-approach-to-machine-learning/">A Data-Driven Approach to Choosing Machine Learning Algorithms</a></li>
</ul>
<h2>Further Reading</h2>
<p>This section provides more resources on the topic if you are looking to go deeper.</p>
<ul>
<li><a href="https://arxiv.org/abs/1708.05070">Data-driven Advice for Applying Machine Learning to Bioinformatics Problems</a></li>
<li><a href="https://github.com/rhiever/sklearn-benchmarks">scikit-learn benchmarks on GitHub</a></li>
<li><a href="https://github.com/EpistasisLab/penn-ml-benchmarks">Penn Machine Learning Benchmarks</a></li>
<li>
<a href="https://crossinvalidation.com/2017/08/22/quantitative-comparison-of-scikit-learns-predictive-models-on-a-large-number-of-machine-learning-datasets-a-good-start/">Quantitative comparison of scikit-learn&rsquo;s predictive models on a large number of machine learning datasets: A good start</a><br>
<a href="https://machinelearningmastery.com/use-random-forest-testing-179-classifiers-121-datasets/">Use Random Forest: Testing 179 Classifiers on 121 Datasets</a>
</li>
</ul>
<h2>Summary</h2>
<p>In this post, you discovered a study and findings from evaluating many machine learning algorithms across a large number of machine learning datasets.</p>
<p>Specifically, you learned:</p>
<ul>
<li>That ensemble tree algorithms perform well across a wide range of datasets.</li>
<li>That it is critical to test a suite of algorithms on a problem as there is no silver bullet algorithm.</li>
<li>That it is critical to test a suite of configurations for a given algorithm as it can result in as much as a 50% improvement on some problems.</li>
</ul>
<p>Do you have any questions?<br>
Ask your questions in the comments below and I will do my best to answer.</p>
<p>The post <a rel="nofollow" href="https://machinelearningmastery.com/start-with-gradient-boosting/">Start With Gradient Boosting, Results from Comparing 13 Algorithms on 165 Datasets</a> appeared first on <a rel="nofollow" href="https://machinelearningmastery.com">Machine Learning Mastery</a>.</p>
<br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning Mastery https://ift.tt/2if3vo8<br>
<br>
March 30, 2018 at 02:00AM
